File: lib/data/registers.ts
import { readFileSync } from "node:fs";
import { join, resolve } from "node:path";

/**
 * Kenyan public registers.
 *
 * None of these publish an API. This module gives each one a uniform
 * adapter backed by a local CSV, so the pipeline is written against a
 * stable interface today and the acquisition method can change without
 * touching anything above.
 *
 * Acquisition, per source, honestly stated:
 *
 *   NCA register of contractors
 *     Portal at nca.go.ke. Search is public, bulk export is not. Section 25
 *     of the NCA Act lets the Authority deregister contractors who default
 *     on annual licence renewal, and a deregistered contractor cannot
 *     re-register under a different name. That makes deregistration a
 *     permanent, checkable negative signal worth capturing. Get bulk access
 *     by asking: the NCA has an institutional interest in the same outcome
 *     you do.
 *
 *   EBK project registration
 *     Launched December 2025. Engineers register a project and receive a
 *     unique number before drawings go to the county. A live site with no
 *     registration number is a real risk flag, and almost nobody is
 *     checking this yet.
 *
 *   Kenya Gazette
 *     Genuinely public at kenyalaw.org. Insolvency notices, winding-up
 *     petitions and liquidator appointments against a developer are the
 *     strongest negative signal available and cost nothing.
 *
 *   Business Registration Service
 *     Company particulars, directors and registered charges via eCitizen.
 *     Paid per search, no bulk. Budget for it on named developers only.
 *
 * Legal note: scraping any of these is a terms question, not just a
 * technical one, and director details are personal data under the Data
 * Protection Act 2019. Register with the ODPC and document a lawful basis
 * before you hold this at scale. That is a week of work now and a dead
 * deal later.
 */

const dataDir = () =>
  process.env.DATA_REGISTERS ?? resolve(process.cwd(), "fixtures", "registers");

export interface ContractorRecord {
  name: string;
  ncaNumber: string;
  category: string;
  status: string; // active | lapsed | deregistered
  licenceExpiry: string | null; // ISO date
  source: "nca";
}

export interface GazetteNotice {
  company: string;
  noticeType: string; // winding-up | insolvency | liquidator | charge
  gazetteDate: string | null; // ISO date
  reference: string;
}

export interface DeveloperCheck {
  name: string;
  contractor: ContractorRecord | null;
  notices: GazetteNotice[];
  projectRegistered: boolean | null;
  flags: string[];
  verdict: string;
}

function rows(filename: string): Array<Record<string, string>> {
  let text: string;
  try {
    text = readFileSync(join(dataDir(), filename), "utf8");
  } catch {
    return [];
  }
  const lines = text.split(/\r?\n/).filter((line) => line.trim().length > 0);
  if (lines.length < 2) return [];
  // The register fixtures contain no quoted fields, so a plain split is a
  // faithful port of csv.DictReader for this data.
  const header = lines[0]!.split(",");
  return lines.slice(1).map((line) => {
    const cells = line.split(",");
    const row: Record<string, string> = {};
    header.forEach((column, i) => {
      row[column] = cells[i] ?? "";
    });
    return row;
  });
}

function isoDate(value: string | undefined): string | null {
  const trimmed = value?.trim() ?? "";
  return /^\d{4}-\d{2}-\d{2}$/.test(trimmed) ? trimmed : null;
}

const NOISE_WORDS = new Set(["ltd", "limited", "company", "co", "and", "the", "&"]);

function norm(name: string): string {
  return name
    .toLowerCase()
    .split(/\s+/)
    .map((word) => word.replace(/^[.,]+|[.,]+$/g, ""))
    .filter((word) => word && !NOISE_WORDS.has(word))
    .join(" ");
}

export function findContractor(name: string): ContractorRecord | null {
  const target = norm(name);
  for (const row of rows("nca_contractors.csv")) {
    if (norm(row.name ?? "") === target) {
      return {
        name: row.name ?? "",
        ncaNumber: row.nca_number ?? "",
        category: row.category ?? "",
        status: row.status ?? "",
        licenceExpiry: isoDate(row.licence_expiry),
        source: "nca",
      };
    }
  }
  return null;
}

export function findNotices(name: string): GazetteNotice[] {
  const target = norm(name);
  const out: GazetteNotice[] = [];
  for (const row of rows("gazette_notices.csv")) {
    if (norm(row.company ?? "") === target) {
      out.push({
        company: row.company ?? "",
        noticeType: row.notice_type ?? "",
        gazetteDate: isoDate(row.gazette_date),
        reference: row.reference ?? "",
      });
    }
  }
  return out.sort((a, b) => (b.gazetteDate ?? "").localeCompare(a.gazetteDate ?? ""));
}

export function projectIsRegistered(projectRef: string): boolean | null {
  const registered = rows("ebk_projects.csv");
  if (!registered.length) return null;
  return registered.some((row) => (row.project_ref ?? "").trim() === projectRef.trim());
}

/**
 * Combine the registers into one verdict.
 *
 * Ordered by severity: an insolvency notice outranks a lapsed licence, and
 * a deregistration outranks both because it is permanent.
 */
export function checkDeveloper(
  name: string,
  projectRef: string | null = null,
  today: string = new Date().toISOString().slice(0, 10),
): DeveloperCheck {
  const contractor = findContractor(name);
  const notices = findNotices(name);
  const registered = projectRef !== null ? projectIsRegistered(projectRef) : null;

  const flags: string[] = [];

  if (contractor === null) {
    flags.push("Not found in the NCA register of contractors");
  } else if (contractor.status === "deregistered") {
    flags.push(
      `Deregistered by the NCA and barred from re-registering (${contractor.ncaNumber})`,
    );
  } else if (contractor.status === "lapsed") {
    flags.push(`NCA licence lapsed (${contractor.ncaNumber})`);
  } else if (contractor.licenceExpiry && contractor.licenceExpiry < today) {
    flags.push(`NCA licence expired ${contractor.licenceExpiry}`);
  }

  for (const notice of notices) {
    if (["winding-up", "insolvency", "liquidator"].includes(notice.noticeType)) {
      flags.push(
        `Gazette ${notice.noticeType} notice dated ${notice.gazetteDate} (${notice.reference})`,
      );
    }
  }

  if (registered === false) {
    flags.push("No EBK project registration found for this project reference");
  }

  const severe = flags.some((flag) =>
    ["deregistered", "winding-up", "insolvency", "liquidator"].some((k) =>
      flag.toLowerCase().includes(k),
    ),
  );

  let verdict: string;
  if (severe) verdict = "do not proceed";
  else if (flags.length) verdict = "proceed with conditions";
  else if (contractor) verdict = "clear";
  else verdict = "unknown";

  return { name, contractor, notices, projectRegistered: registered, flags, verdict };
}

File: offplan-escrow-mvp/datasources/registers.py
"""
Kenyan public registers.

None of these publish an API. This module gives each one a uniform adapter
backed by a local CSV, so the pipeline is written against a stable interface
today and the acquisition method can change without touching anything above.

Acquisition, per source, honestly stated:

  NCA register of contractors
    Portal at nca.go.ke. Search is public, bulk export is not. Section 25 of
    the NCA Act lets the Authority deregister contractors who default on
    annual licence renewal, and a deregistered contractor cannot re-register
    under a different name. That makes deregistration a permanent, checkable
    negative signal worth capturing. Get bulk access by asking: the NCA has an
    institutional interest in the same outcome you do.

  EBK project registration
    Launched December 2025. Engineers register a project and receive a unique
    number before drawings go to the county. A live site with no registration
    number is a real risk flag, and almost nobody is checking this yet.

  Kenya Gazette
    Genuinely public at kenyalaw.org. Insolvency notices, winding-up petitions
    and liquidator appointments against a developer are the strongest negative
    signal available and cost nothing.

  Business Registration Service
    Company particulars, directors and registered charges via eCitizen. Paid
    per search, no bulk. Budget for it on named developers only.

Legal note: scraping any of these is a terms question, not just a technical
one, and director details are personal data under the Data Protection Act
2019. Register with the ODPC and document a lawful basis before you hold this
at scale. That is a week of work now and a dead deal later.
"""

from __future__ import annotations

import csv
from dataclasses import dataclass, field
from datetime import date, datetime
from pathlib import Path

DATA_DIR = Path(__file__).resolve().parent.parent / "fixtures" / "registers"


@dataclass
class ContractorRecord:
    name: str
    nca_number: str
    category: str
    status: str  # active | lapsed | deregistered
    licence_expiry: date | None
    source: str = "nca"


@dataclass
class GazetteNotice:
    company: str
    notice_type: str  # winding-up | insolvency | liquidator | charge
    gazette_date: date
    reference: str


@dataclass
class DeveloperCheck:
    name: str
    contractor: ContractorRecord | None
    notices: list[GazetteNotice] = field(default_factory=list)
    project_registered: bool | None = None
    flags: list[str] = field(default_factory=list)
    verdict: str = "unknown"


def _rows(filename: str) -> list[dict]:
    path = DATA_DIR / filename
    if not path.exists():
        return []
    with path.open(newline="", encoding="utf-8") as fh:
        return list(csv.DictReader(fh))


def _date(value: str) -> date | None:
    try:
        return datetime.strptime(value.strip(), "%Y-%m-%d").date()
    except (ValueError, AttributeError):
        return None


def _norm(name: str) -> str:
    drop = {"ltd", "limited", "company", "co", "and", "the", "&"}
    words = [w.strip(".,") for w in name.lower().split()]
    return " ".join(w for w in words if w and w not in drop)


def find_contractor(name: str) -> ContractorRecord | None:
    target = _norm(name)
    for row in _rows("nca_contractors.csv"):
        if _norm(row["name"]) == target:
            return ContractorRecord(
                name=row["name"],
                nca_number=row["nca_number"],
                category=row["category"],
                status=row["status"],
                licence_expiry=_date(row["licence_expiry"]),
            )
    return None


def find_notices(name: str) -> list[GazetteNotice]:
    target = _norm(name)
    out = []
    for row in _rows("gazette_notices.csv"):
        if _norm(row["company"]) == target:
            out.append(GazetteNotice(
                company=row["company"],
                notice_type=row["notice_type"],
                gazette_date=_date(row["gazette_date"]),
                reference=row["reference"],
            ))
    return sorted(out, key=lambda n: n.gazette_date or date.min, reverse=True)


def project_is_registered(project_ref: str) -> bool | None:
    rows = _rows("ebk_projects.csv")
    if not rows:
        return None
    return any(r["project_ref"].strip() == project_ref.strip() for r in rows)


def check_developer(name: str, project_ref: str | None = None, today: date | None = None) -> DeveloperCheck:
    """
    Combine the registers into one verdict.

    Ordered by severity: an insolvency notice outranks a lapsed licence, and a
    deregistration outranks both because it is permanent.
    """
    today = today or date.today()
    contractor = find_contractor(name)
    notices = find_notices(name)
    registered = project_is_registered(project_ref) if project_ref else None

    flags: list[str] = []

    if contractor is None:
        flags.append("Not found in the NCA register of contractors")
    elif contractor.status == "deregistered":
        flags.append(f"Deregistered by the NCA and barred from re-registering ({contractor.nca_number})")
    elif contractor.status == "lapsed":
        flags.append(f"NCA licence lapsed ({contractor.nca_number})")
    elif contractor.licence_expiry and contractor.licence_expiry < today:
        flags.append(f"NCA licence expired {contractor.licence_expiry}")

    for n in notices:
        if n.notice_type in ("winding-up", "insolvency", "liquidator"):
            flags.append(f"Gazette {n.notice_type} notice dated {n.gazette_date} ({n.reference})")

    if registered is False:
        flags.append("No EBK project registration found for this project reference")

    severe = any(
        k in f.lower()
        for f in flags
        for k in ("deregistered", "winding-up", "insolvency", "liquidator")
    )
    if severe:
        verdict = "do not proceed"
    elif flags:
        verdict = "proceed with conditions"
    elif contractor:
        verdict = "clear"
    else:
        verdict = "unknown"

    return DeveloperCheck(
        name=name,
        contractor=contractor,
        notices=notices,
        project_registered=registered,
        flags=flags,
        verdict=verdict,
    )
